### DISCUSSION TITLE
Boolean attributes

### DISCUSSION THREAD

**Discussion by goertzenator** [chrisdone/lucid#159] [Ideas]
HTML boolean attributes are awkward to deal with because their presence or absence indicates their value rather than a boolean parameter.  When working with these attributes in Lucid2 one is compelled to do things like this:

```
i1 = input_ ([type_ "checkbox", name_ "check1" ] <> if doChecked then [checked_] else [])
i2 = input_ ([type_ "checkbox", name_ "check2" ] <> ([checked_ | doChecked]))
i3 = input_ [type_ "checkbox", name_ "check3", if doChecked then checked_ else mempty]
i4 = input_ [type_ "checkbox", name_ "check4", bool mempty checked_ doChecked]
```

React and Alpine.js (and likely many others) paper over that by making boolean attributes work like a parameterized attribute taking a bool, which works out nicely:
```
i5 = input_ [type_ "checkbox", name_ "check5", checked_' doChecked]
```

I'd like to suggest that parameterized boolean attributes be added to Lucid2:

```
checked_' :: Bool -> Attributes
checked_' = bool mempty checked_

autofocus_' :: Bool -> Attributes
autofocus_' = bool mempty autofocus_

disabled_' :: Bool -> Attributes
disabled_' = bool mempty (makeAttributes "disabled" mempty)
```

One other observation: The existing `disabled_` attribute takes a `Text` parameter.  This is tolerated by HTML5, [but does nothing](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML).  It should probably be removed, or if it needs to be preserved for compatibility, at least document that it functionally does nothing.

Thoughts?

**Comment by chrisdone** [comment]
Reasonable topic.

Taking a Bool is simple and an easy addition. Although I’d prefer some other naming convention than the prime. Maybe “checkedIf_” or something like that.

I won’t introduce breaking changes but very happy to add docs regarding “no-op” attribute values, perhaps pointing to the better option.

Haven’t been at the personal laptop keyboard for a few days, sorry for the slow reply.

On Wed, Oct 30, 2024, at 2:11 PM, Daniel Goertzen wrote:
> 
> 
> HTML boolean attributes are awkward to deal with because their presence or absence indicates their value rather than a boolean parameter. When working with these attributes in Lucid2 one is compelled to do things like this:
> 
> `i1 = input_ ([type_ "checkbox", name_ "check1" ] <> if doChecked then [checked_] else [])
> i2 = input_ ([type_ "checkbox", name_ "check2" ] <> ([checked_ | doChecked]))
> i3 = input_ [type_ "checkbox", name_ "check3", if doChecked then checked_ else mempty]
> i4 = input_ [type_ "checkbox", name_ "check4", bool mempty checked_ doChecked]
`
> React and Alpine.js (and likely many others) paper over that by making boolean attributes work like a parameterized attribute taking a bool, which works out nicely:
> 
> `i5 = input_ [type_ "checkbox", name_ "check5", checked_' doChecked]
`
> I'd like to suggest that parameterized boolean attributes be added to Lucid2:
> 
> `checked_' :: Bool -> Attributes
> checked_' = bool mempty checked_
> 
> autofocus_' :: Bool -> Attributes
> autofocus_' = bool mempty autofocus_
> 
> disabled_' :: Bool -> Attributes
> disabled_' = bool mempty (makeAttributes "disabled" mempty)
`
